### ISSUE TITLE
Name Service endpoint throws 404 error for name/address not found

### ISSUE THREAD

**Issue by tboeckmann** [mailchain/mailchain#242]
**Describe the bug**
When calling endpoints with a valid name/address but one that does not have a resolvable ENS name/address, the API response is a 404.

* http://127.0.0.1:8080/api/nameservice/address/{address}/resolve?protocol=ethereum&network=ropsten
* http://127.0.0.1:8080/api/nameservice/name/{name}/resolve?protocol=ethereum&network=ropsten

This is a arguably correct response for no name existing, but it leads to degraded developer/user experience as:
[1] The 'not found' error needs to be handled as part of the application flow, as an expected result.
[2] The error is propagated to the browser console and looks, well, ugly 👹

**To Reproduce**
1. Go to 'https://inbox.mailchain.xyz/#/' with mailchain client running with addresses that are not configured for reverse resolution in ENS 
2. Open the javascript console
3. See error

**What version of Mailchain are you using ?**
`$ mailchain version`
`Version: v0.0.31`

**Expected behavior**
Something more graceful - open to ideas that help the developer handle errors better

**Comment by robdefeo** [comment]
A not found resolution can be interpreted as a 2XX error being that the resource is available and the response is this name/address can not be resolved. An appropriate response needs to be defined to pair with the changed response code.

**Comment by robdefeo** [comment]
Resolve Name returns a response that looks like 
```
{
  "address": "0x4ad2b251246aafc2f3bdf3b690de3bf906622c51"
}
```
Resolve Address returns 
```
{
  "name": "mailchain.eth"
}
```
Whats a good response? 
- empty values - is unclear what empty really means
- adding a status code - 200 for OK 404 for not resolvable - This would conflict with 200 status in not found situatoin
- other

**Comment by tboeckmann** [comment]
Empty string seems fine because we're only concerned with a positive response rather than worrying about why it's not returning a value.

```
{
  "address": ""
}

```

```
{
  "name": ""
}
```

**Comment by robdefeo** [comment]
For field value this is ok, but lacks clarity on what happened. What about using the DNS return codes as per https://tools.ietf.org/html/rfc2929 ? 

| RCODE  | Name |    Description |                        Reference |
|----|---|---| --- |
 |   0|    NoError  | No Error              |             [RFC 1035] |
 |   1  |  FormErr  | Format Error        |               [RFC 1035] |
 |   2   | ServFail  |Server Failure        |             [RFC 1035] |
 |   3   | NXDomain | Non-Existent Domain |               [RFC 1035] |
 |   4   | NotImp  |  Not Implemented             |       [RFC 1035] |
 |   5   | Refused  | Query Refused                 |     [RFC 1035] |

**Comment by nikosk686** [comment]
I could apply a solution on the handlers functions.
By performing a special handling for 'not found resolution'
cases. But I am not sure if this would be correct in general.
Maybe the mapping of 404 to 2xx should happen before the 
handlers function?

**Comment by robdefeo** [comment]
@nikosk686 you raise a valid point, while its easy to catch the `IsNotFoundError` i think there is a better solution to this that involves replacing the errors in https://github.com/mailchain/mailchain/blob/master/nameservice/errors.go with errors that correspond to the table in https://github.com/mailchain/mailchain/issues/242#issuecomment-529144563 this would mean then the http handler contains less logic (a good thing). I like the solution you are proposing a lot :)

**Comment by nikosk686** [comment]
Hi @robdefeo , thanks for your input.
I totally agree that the handlers should contain less logic but I am not sure where the replacement of the error could take place, yet. I ll try to apply the changes where I consider they properly fit and we can further discuss upon the pull request, if better solution can be applied.

**Comment by robdefeo** [comment]
Happy to help in the PR but before you get too far into, I'd look at replacing the errors as we discussed previously. Then using the case statement here https://github.com/mailchain/mailchain/blob/6f6b1738cc536c2b3ff42a3b6f3e919a9ee7933e/nameservice/ens/errors.go to return one of the new errors.  Then is should be a matter of replacing https://github.com/mailchain/mailchain/blob/362ff8b90a738d5a43d06e8c08b8149000cc68ec/cmd/mailchain/internal/http/handlers/resolve_name.go#L63-L74 with something that checks if the error is a "known" error. The same in https://github.com/mailchain/mailchain/blob/362ff8b90a738d5a43d06e8c08b8149000cc68ec/cmd/mailchain/internal/http/handlers/resolve_address.go#L63-L74

**Comment by robdefeo** [comment]
Of course @nikosk686 as you get in to solving the issue your likely to find a more optimal solution, I'd consider this a possible starting point rather than the exact solution. As always when you work on the issue you find out more as you pull the threads, just like you did the first time 😄

**Comment by nikosk686** [comment]
Hi @robdefeo,
I ve created a pr for this, according to (what I believe was) your input above.

Some things to consider, and not applied in this fix, are related to

- Making the mapping of 404 to the rfc1035 specified error optional, either with a config param or a request param sent from the client, i.e. a client can decide to receive the original 404 error.
- There is a lot of common code in resolve_name and resolve_address handlers and tehir tests, should be refactored.
- I am not sure if related changes should be done on nameservice.ens package.

**Comment by robdefeo** [comment]
Hi @nikosk686, thanks for the PR, I have left comments in the review of #266 I like your approach and with some tweaks it will solve the problem in an elegant way. For your points above.
+ Lets keep this out of scope, changing the REST response behaviour based on a setting feels uncomfortable. It would be unclear whats happening for many developers and would require knowing the value of a setting to debug. I think its better to return 200's in the case of domain is valid it just can't resolve the name.
+ Agreed, but happy to keep seperate especially in case of test suits. Just incase things diverge tests still ensure behaviour as expected.
+ Yes, I left comments about this in the PR.
